core/src/main/scala/kafka/admin/ReassignPartitionsCommand.scala

```diff
@@ -16,233 +16,1555 @@
+   * @param topics          The topics to get information about.
+   * @return                A map from partitions to broker assignments.
+   *                        If any topic can't be found, an exception will be thrown.
+   */
+  def getReplicaAssignmentForTopics(adminClient: Admin,
+                                    topics: Seq[String])
+                                    : Map[TopicPartition, Seq[Int]] = {
+    adminClient.describeTopics(topics.asJava).all().get().asScala.flatMap {
+      case (topicName, topicDescription) => {
+        topicDescription.partitions().asScala.map {
+          info => (new TopicPartition(topicName, info.partition()),
+            info.replicas().asScala.map(_.id()))
+        }
+      }
+    }
   }
 
-  def executeAssignment(zkClient: KafkaZkClient, adminClientOpt: Option[Admin], reassignmentJsonString: String, throttle: Throttle, timeoutMs: Long = 10000L): Unit = {
-    val (partitionAssignment, replicaAssignment) = parseAndValidate(zkClient, reassignmentJsonString)
+  /**
+   * Get the current replica assignments for some partitions.
+   *
+   * @param adminClient     The AdminClient to use.
+   * @param partitions      The partitions to get information about.
+   * @return                A map from partitions to broker assignments.
+   *                        If any topic can't be found, an exception will be thrown.
+   */
+  def getReplicaAssignmentForPartitions(adminClient: Admin,
+                                        partitions: Set[TopicPartition])
+                                        : Map[TopicPartition, Seq[Int]] = {
+    val topics = new mutable.HashSet[String]()
+    partitions.foreach(topics += _.topic)
+    adminClient.describeTopics(topics.asJava).all().get().asScala.flatMap {
+      case (topicName, topicDescription) => {
+        topicDescription.partitions().asScala.flatMap {
+          info => if (partitions.contains(new TopicPartition(topicName, info.partition()))) {
+            Some(new TopicPartition(topicName, info.partition()),
+                info.replicas().asScala.map(_.id()))
+          } else {
+            None
+          }
+        }
+      }
+    }
+  }
+
+  /**
+   * Find the rack information for some brokers.
+   *
+   * @param adminClient         The AdminClient object.
+   * @param brokers             The brokers to gather metadata about.
+   * @param enableRackAwareness True if we should return rack information, and throw an
+   *                            exception if it is inconsistent.
+   *
+   * @return                    The metadata for each broker that was found.
+   *                            Brokers that were not found will be omitted.
+   */
+  def getBrokerMetadata(adminClient: Admin,
+                        brokers: Seq[Int],
+                        enableRackAwareness: Boolean): Seq[BrokerMetadata] = {
+    val brokerSet = brokers.toSet
+    val results = adminClient.describeCluster().nodes().get().asScala.flatMap(
```

Comment:
nit: using a `filter` followed with a `map` in might be a little more readable here